core/camel-support/src/main/java/org/apache/camel/support/InputStreamIterator.java

```diff
@@ -65,11 +67,18 @@ public int available() throws IOException {
 
     private InputStream nextChunk() throws IOException {
         if (it.hasNext()) {
+            Object next = null;
             try {
-                byte[] buf = converter.mandatoryConvertTo(byte[].class, it.next());
+                next = it.next();
+                byte[] buf = converter.mandatoryConvertTo(byte[].class, next);
                 return new ByteArrayInputStream(buf);
             } catch (NoTypeConversionAvailableException e) {
                 throw new IOException(e);
+            } finally {
+                // ensure the original object is closed if it was closeable
+                if (next != null && next instanceof Closeable closeableNext) {
```

Comment:
This is also wrong as it.next can be part of some object that should only be closed when its fully done.